Fixes #8797 - Separate style from the HTML in emails

amirfefer opened the pull request:
By using `roadie-rails` gem  style can be separated from email html foramt. email style was separated to 'email.css' file.

amirfefer committed:
Fixes #8797 - Separate style from the HTML in emails

By using roadie-rails gem (1.1) style can be separated from email
html foramt, and absolute paths can be used.
style was separated to 'email.css'

domcleal commented on Gemfile:
```diff
@@ -29,6 +29,7 @@ gem 'i18n', '~> 0.6.4'
 gem 'rails-i18n', '~> 3.0.0'
 gem 'turbolinks', '~> 2.5'
 gem 'logging', '>= 1.8.0', '< 3.0.0'
+gem 'roadie', '~> 2.4'
```
The current version is 3.0.5, with 3.1.0 in rc.  Could we start with that, or is there something blocking use of 3.x?

domcleal replied:
Looks like roadie-rails is the right gem to be using to use roadie 3.x with Rails.

domcleal commented:
I'm going to leave this largely unreviewed until the 3.x comment is resolved, since I suspect there would be more changes.

There are other features of the gem that could be used, such as the conversion for absolute paths.

Lastly, in a production environment, when is the CSS used?  Presumably it's at runtime, so the file under app/assets/stylesheets/ is going to need to be present - but not precompiled, right?

amirfefer commented:
The gem changed to `roadie-rails`. 
@domcleal  About your last question, According to roadie doc:

> ...when generating your email  Roadie will look for the precompiled files on your filesystem, or by asking the asset pipeline to compile the files for you if it cannot be found.

domcleal commented on Gemfile:
```diff
@@ -38,6 +38,7 @@ end
 gem 'activerecord-session_store', '~> 0.1.1'
 gem 'rails-observers', '~> 0.1'
 gem 'protected_attributes', '~> 1.1.1'
+gem 'roadie-rails'
```
This should have version pinning based on the update policy of the gem, and the minimum version that the functionality required is in.

domcleal commented on config/initializers/assets.rb:
```diff
@@ -102,4 +102,7 @@
   end
+  #Absulote static URL for mailer
+  #url = URI.parse(Setting[:foreman_url])
+  #qqconfig.roadie.url_options = {:host => url.host, :port => url.port, :protocol => url.scheme}
```
This shouldn't be here.

domcleal replied:
Well, or should it be moved from development.rb so it applies to all environments?

What I mean to say above is - there shouldn't be commented out code in a proposed patch.

domcleal commented on config/environments/development.rb:
```diff
@@ -47,6 +47,9 @@
     Bullet.console = true
     Bullet.rails_logger = true
     Bullet.add_footer = true
+    #Absulote static URL for mailer
```
"Absolute"

domcleal commented:
Please look at the integration tests, they're failing.

<pre>
ActionController::RoutingError: No route matches [GET] "/var/lib/workspace/workspace/test_develop_pr_core@2/database/postgresql/label/fast/ruby/1.9.3/app/assets/stylesheets/email.css"
     (ActionController::RoutingError)
/usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pr_core-2/gems/actionpack-4.1.5/lib/action_dispatch/middleware/debug_exceptions.rb:21    
</pre>

> ...when generating your email Roadie will look for the precompiled files on your filesystem, or by asking the asset pipeline to compile the files for you if it cannot be found.

The file(s) will need to be added to the precompilation list in production.rb, assuming that works with roadie as the asset pipeline won't be available in a production environment.

amirfefer commented:
@domcleal  I fixed your comments and rebased.  Instead of using the Rails.config for  absolute path, I used roadie_options in `application mailer`, for dynamic support.  

EDIT: It seems that some integration tests have failed because capybara without actual effect. as you can see the test I added ( in `ApplicationMailer`) which examines whether css file can be merged into html email, passed successfully .

domcleal commented:
> It seems that some integration tests have failed because capybara without actual effect. as you can see the test I added ( in ApplicationMailer) which examines whether css file can be merged into html email, passed successfully .

The tests still need to be fixed.

amirfefer commented on app/assets/stylesheets/application.scss:
```diff
@@ -11,7 +11,7 @@
 @import "patternfly";
 @import "mixin";
-@import "/**/*";
+@import "/*";
```
`@import "/**/*"` imports all files in dir including subdir.  In that case  `email.css` is included to `application.scss`, and caused the integration errors.  just roadie uses it, hence I don't see any point to import it to `application.scss`.
`@import "/*"` imports all files in dir except subdir files.  So I created a new subdir in stylesheets for un-imported files.

domcleal commented on Gemfile:
```diff
@@ -38,6 +38,7 @@ gem 'activerecord-session_store', '~> 0.1.1'
 gem 'protected_attributes', '~> 1.1.1'
 gem 'sprockets', '~> 3'
 gem 'sprockets-rails', '>= 2.3.3', '< 3'
+gem 'roadie-rails', '~> 1.1.0'
```
Should be `~> 1.1` as the gem follows semver.

domcleal commented on app/assets/stylesheets/unimported/email.css:
```diff
@@ -0,0 +1,87 @@
+}
+
+.dashboard-td {
+    font-size: 125%;
```
Indentation is inconsistent throughout, prefer two spaces.

domcleal commented on app/assets/stylesheets/unimported/email.css:
```diff
@@ -0,0 +1,87 @@
+  font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif;
+  color: #3f3f3f;
+  background-color: #f1f1f1;
+  padding: 10px 24px
```
missing `;`, ditto in other places

domcleal commented on app/views/audit_mailer/summary.html.erb:
```diff
@@ -1,20 +1,23 @@
         %>
         <b>
-          <%= link_to(audit.username, audits_url(:host => @url.host, :port => @url.port, :protocol => @url.scheme, :search => user_search_param)) %>
+          <%= link_to(audit.username, audits_path(:search => user_search_param)) %>
```
Rails 4.2 logs deprecation warnings for using _path in a mailer, does Roadie override this?

domcleal commented on app/assets/stylesheets/unimported/email.css:
```diff
@@ -0,0 +1,87 @@
+  border: 1px solid #e1e2e3
+}
+
+.email-td {
```
Move next to the headers/table classes?

Why use the `email-` prefix here, but not on other classes?  Would it make sense to either drop the prefix or apply the class only to the `table` element and then define `.email-table < td` or similar?

domcleal commented on app/views/audit_mailer/summary.html.erb:
```diff
@@ -1,20 +1,23 @@
-<body style="font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif; color: #3f3f3f; background-color: #f1f1f1; padding: 10px 24px">
-<h2 style="font-weight: normal; text-transform: uppercase; font-size: 120%;"><%= _("<b>Foreman</b> audit summary").html_safe %></h2>
-<ul style="list-style-type: none;">
+<head>
```
To prevent duplicating this into each mailer, add a layout that's used for everything containing the standard html/head/stylesheet/body headers/footers.

domcleal commented:
Thanks for the update.  If you could please tidy the stylesheet and classes as much as possible, that'd be appreciated - it's very thrown together and inconsistent, from where it's been copied from different places.

amirfefer commented on app/views/application_mailer/_header.html.erb:
```diff
@@ -0,0 +1,7 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
```
@domcleal Is that the correct path for the partials?

domcleal replied:
My suggestion was to use layouts rather than partials for these - is that possible?  At least you'd not have two parts of the same piece of HTML split across files.

domcleal commented:
This PR isn't currently mergeable, please rebase it.

domcleal commented on app/views/layouts/application_mailer.html.erb:
```diff
@@ -0,0 +1,11 @@
+    <%= yield :head %>
+  </head>
+  <body>
+    <%= yield :body %>
```
Use a plain `yield` here with no name, then you don't need to put `content_for :body` into every single view, same as with a regular web UI layout.

domcleal commented on app/views/user_mailer/tester.html.erb:
```diff
@@ -1,4 +1,4 @@
-<h2 style="font-weight: normal; text-transform: uppercase; font-size: 120%;"><%= _("<b>Foreman</b> test email").html_safe %></h2>
-<h4><%= _("This is a test message to confirm that Foreman's email configuration is working.") %></h4>
-</body>
+<% content_for :body do %>
```
These can all then be removed, and unintend the content by two spaces.

domcleal commented on app/views/layouts/application_mailer.html.erb:
```diff
@@ -0,0 +1,11 @@
+  </head>
+  <body>
+    <%= yield :body %>
+   </body>
```
indentation is off, two spaces

domcleal commented:
It needs another rebase due to an audit mailer change.

amirfefer commented:
@domcleal PR is fixed and rebased

domcleal commented on app/views/audit_mailer/summary.html.erb:
```diff
@@ -1,20 +1,17 @@
+           user_search_param = login ? "user = #{login}" : %Q{username = "#{audit.username}"} %>
         <b>
-          <%= link_to(audit.username, audits_url(:host => @url.host, :port => @url.port, :protocol => @url.scheme, :search => user_search_param)) %>
+          <%= link_to(audit.username, audits_url(:search => user_search_param, :only_path => true)) %>
```
Rather than `only_path => true`, shouldn't this be `audits_path`?  (Assuming roadie converts the path to a URL.)

domcleal commented on app/views/host_mailer/error_state.html.erb:
```diff
@@ -1,14 +1,6 @@
+<div class="dashboard">
+  <%= render :partial => 'error_states', :locals => {:logs => @report.logs} %>
+  <br>
+  <%= link_to _("For more information"), config_report_url(@report, :host => @url.host, :port => @url.port, :protocol => @url.scheme) %>
```
`@url` references should be removed here, and then perhaps use config_report_path per above comment?

amirfefer commented:
@domcleal  Links have been changed to `_path` instead of `_url` (roadie converts path to url).

domcleal commented on app/assets/stylesheets/unimported/email.css:
```diff
@@ -0,0 +1,91 @@
+  margin: 5px 0px 10px;
+}
+
+p.dashboard {
```
`p td.dashboard`, or just `p .dashboard` perhaps?  Saves re-specifying class on every `<p>` element.

amirfefer replied:
@domcleal  what about `td.dashboard p` ?   p element is a child of td element with `dashboard` class.

domcleal commented:
There's a small issue with the CSS in the Puppet summary reports, as a table border has appeared around some elements:

Before:
![screenshot from 2016-05-24 13-48-26](https://cloud.githubusercontent.com/assets/482089/15504403/aa83ae18-21b6-11e6-826b-6d55c0cc29d8.png)

After:
![screenshot from 2016-05-24 13-48-21](https://cloud.githubusercontent.com/assets/482089/15504407/b0424148-21b6-11e6-8b5a-18cea7d0ac4f.png)

amirfefer commented:
@domcleal the css has been fixed.

domcleal commented:
Merged as a1506ca, thanks @amirfefer.
